src/TSTransformer/util/checkVariableHoist.ts

```diff
@@ -0,0 +1,39 @@
+import { getOrSetDefault } from "Shared/util/getOrSetDefault";
+import { TransformState } from "TSTransformer/classes/TransformState";
+import { getAncestor, isAncestorOf } from "TSTransformer/util/traversal";
+import ts from "typescript";
+
+export function checkVariableHoist(state: TransformState, node: ts.Identifier, symbol: ts.Symbol) {
+	if (state.isHoisted.get(symbol) !== undefined) {
+		return;
+	}
+
+	const statement = getAncestor(node, ts.isStatement);
+	if (!statement) {
+		return;
+	}
+
+	const caseClause = statement.parent;
+	if (!ts.isCaseClause(caseClause)) {
+		return;
+	}
+	const caseBlock = caseClause.parent;
+
+	const isUsedOutsideOfCaseClause =
+		ts.FindAllReferences.Core.eachSymbolReferenceInFile(
+			node,
+			state.typeChecker,
+			node.getSourceFile(),
+			token => {
+				if (!isAncestorOf(caseClause, token)) {
+					return true;
+				}
+			},
+			caseBlock,
+		) === true;
+
+	if (isUsedOutsideOfCaseClause) {
+		getOrSetDefault(state.hoistsByStatement, statement.parent, () => new Array<ts.Identifier>()).push(node);
+		state.isHoisted.set(symbol, true);
+	}
+}
```

Comment:
Are there any actual changes here, or just separating it to util? Breaking it down into smaller commits can help with review diffs.

Reply:
Just putting it into a util file